src/javaharness/java/arcs/webimpl/PortablePromiseJsImpl.java

```diff
@@ -0,0 +1,29 @@
+package arcs.webimpl;
+
+import arcs.api.PortablePromise;
+import elemental2.promise.Promise;
+import elemental2.promise.Promise.PromiseExecutorCallbackFn;
+import elemental2.promise.IThenable.ThenOnFulfilledCallbackFn;
+import elemental2.promise.IThenable;
+import java.util.function.Consumer;
+
+
+class PortablePromiseJsImpl<T> implements PortablePromise<T> {
+  private final PortablePromise.PortablePromiseExecutor<T> executor;
+  private Promise<T> promise;
+
+  public PortablePromiseJsImpl(PortablePromise.PortablePromiseExecutor<T> executor) {
+    this.promise = new Promise((resolve, reject) -> {
+      executor.doInvoke((T value) -> resolve.onInvoke(value),
+                        (Object error) -> reject.onInvoke(error));
+    });
+    this.executor = executor;
+  }
+
+  @Override
+  public PortablePromise<T> then(Consumer<T> onFulfillment) {
```

Comment:
I think that this method signature may be wrong, since a chain of promise handlers can keep changing types, e.g.

promise.then(x -> x * 2).then(y -> "Hello " + y).then(...)


So this needs to be public <R> PortablePromise<R> then(Function<T, R>)

That is, you pass in a function that maps T to R.  Or you could add this as an overload, so that Consumer<T> is for void returning functions.

Reply:
Actually the Elemental2 promise doesn't seem to expose the appropriate API so I'm not sure chaining will work. You can land as is, and we can refactor afterwards.

Reply:
elemental2 has 
```
 public native <V> Promise<V> then(
      ThenOnFulfilledCallbackFn<? super T, ? extends V> onFulfilled,
      ThenOnRejectedCallbackFn<? extends V> onRejected);
```
Isn't that the equivalent of Function<T, R> you suggested above? So that the new interface would be
``` public PortablePromise<R> then(Function<T, R> onFulfillment) {...} ```?
(with an overloading for just Consumer as in `public PortablePromise<void> then(Consumer<T> onFulfillment) {...}`?

But i'm happy to land as is and figure it out later, so that i can make progress on the PEC work. Shall I just remove the return value altogether, WDYT?

Reply:
The problem is ThenOnFulfilledCallbackFn returns only IThenable, not primitive values, so something like promise.then(x -> x * 2) is impossible, since then .then() argument must return an object that implements IThenable, so you'd need to do promise.then(x -> Promise.resolve(x*2))

I pinged the J2CL guys to ask, but it seems to me the underlying full browser API isn't exposed. We can work around it, but maybe land this first and then we can tweak the API.

Reply:
I guess what I'm saying it, you could have PortablePromise.then(Function<T, R> func) in the portable API, underneath, it translates this into calling promise.then(x -> Promise.resolve(func.apply(x))), because the Elemental2 library requires the fulfilled function to return an IThenable (Promise implements IThenable)

Or you could leave this, and I'll fix up the APIs after you commit.

Reply:
sounds good, thanks, landing as is.